sway/server.c

```diff
@@ -101,3 +124,289 @@ void server_run(struct sway_server *server) {
+		wl_list_remove(&subbackend->backend_destroy.link);
+	}
+	wl_list_remove(&subbackend->link);
+	// free(name)?
+	free(subbackend);
+}
+
+struct sway_subbackend *sway_server_get_subbackend(struct sway_server *server,
+		char *name) {
+	struct sway_subbackend *subbackend = NULL;
+	wl_list_for_each(subbackend, &server->subbackends, link) {
+		if (strcasecmp(subbackend->name, name) == 0) {
+			return subbackend;
+		}
+	}
+
+	return NULL;
+}
+
+struct sway_subbackend *sway_subbackend_create(enum sway_subbackend_type type,
+		char *name) {
+	struct sway_subbackend *subbackend =
+		calloc(1, sizeof(struct sway_subbackend));
+	if (subbackend == NULL) {
+		wlr_log(L_ERROR, "could not allocate subbackend");
+		return NULL;
+	}
+
+	if (name == NULL) {
+		// TODO: figure out what to call the backend based on what the backend
+		// type is and how many other backends are configured of that type
+		// (<type>-<num>).
+	} else {
+		subbackend->name = name;
+	}
+
+	subbackend->type = type;
+	wl_list_init(&subbackend->outputs);
+	wl_list_init(&subbackend->inputs);
+	wl_list_init(&subbackend->link);
+
+	return subbackend;
+}
+
+static void handle_subbackend_backend_destroy(struct wl_listener *listener,
+		void *data) {
+	struct sway_subbackend *subbackend =
+		wl_container_of(listener, subbackend, backend_destroy);
+	sway_subbackend_destroy(subbackend);
+}
+
+static struct wlr_backend *wayland_backend_create(struct sway_server *server) {
+	wlr_log(L_DEBUG, "TODO: create wayland backend");
+	return NULL;
+	/*
+	struct wlr_backend *backend =
+		wlr_wl_backend_create(server->wl_display);
+	wlr_wl_output_create(backend);
+	return backend;
+	*/
+}
+
+static struct wlr_backend *x11_backend_create(struct sway_server *server) {
+	wlr_log(L_DEBUG, "TODO: create x11 backend");
+	return NULL;
+}
+
+static struct wlr_backend *headless_backend_create(
+		struct sway_server *server) {
+	struct wlr_backend *backend =
+		wlr_headless_backend_create(server->wl_display);
+	return backend;
+}
+
+static struct wlr_backend *drm_backend_create(struct sway_server *server) {
+	wlr_log(L_DEBUG, "TODO: create drm backend");
+	return NULL;
+}
+
+void sway_server_add_subbackend(struct sway_server *server,
+		struct sway_subbackend *subbackend) {
+	if (sway_server_get_subbackend(server, subbackend->name)) {
+		wlr_log(L_ERROR, "cannot add subbackend '%s': already exists",
+			subbackend->name);
+		sway_subbackend_destroy(subbackend);
+		return;
+	}
+
+	struct wlr_backend *backend = NULL;
+
+	switch (subbackend->type) {
+	case SWAY_SUBBACKEND_WAYLAND:
+		backend = wayland_backend_create(server);
+		break;
+	case SWAY_SUBBACKEND_X11:
+		backend = x11_backend_create(server);
+		break;
+	case SWAY_SUBBACKEND_DRM:
+		backend = drm_backend_create(server);
+		break;
+	case SWAY_SUBBACKEND_HEADLESS:
+		backend = headless_backend_create(server);
+		break;
+	}
+
+	if (backend == NULL) {
+		wlr_log(L_ERROR, "could not create subbackend '%s'", subbackend->name);
+		sway_subbackend_destroy(subbackend);
```

Comment:
I don't like "subbackend" everywhere, just call it backend (or backend_config maybe for this type)